### DISCUSSION TITLE
CORS issue in cefsharp.winforms when not in Chrome

### DISCUSSION THREAD

**Discussion by smillsoper** [cefsharp/CefSharp#4263] [General]
I came across a CORS issue in CefSharp.Winforms when in full-blown Chrome, I don't get the CORS errors.

I had not done any javascript/jquery injection and still got the errors.

To reproduce, navigate to https://scoopfit.com and select a product to order, and go through the checkout process up until the payment method options.  Select Afterpay and click pay now (don't worry, we're not actually ordering).

Once that page loads, open developer tools and go to the console tab.  

If the above was done in Chrome, there will be no CORS errors although there are some other errors regarding inline scripts and resource load errors (but that's just Shopify - I expect to see that).

If the above was loaded in CefSharp, there are a number of CORS errors.

I have attempted to initialize chrome with the disable-web-security command line option, but that does not disable CORS.  Is there something I can do to circumvent this issue?  This is the first time I have come across this issue.

I'm using the latest version of CefSharp.WinForms (v105); and also had the same issue on version 99.  I updated to 105 from 99 to see if it would correct the issue, but it did not.

I have also attempted using the AddCrossOriginWhitelistEntry() method still to no avail (although I expected that not to work as I have seen that this method needs to be corrected in another discussion).

This one has me totally stumped just for the fact that I'm not making any requests other than just loading the site as in a full Chrome browser and not getting the same result.  With those CORS errors, even if you populate the phone number or email address, the submit button will not work and you will see 'too many requests' errors when you click that button.

any help would be appreciated.

**Comment by amaitland** [comment]
You'll need to test with the [CEF Sample Application](https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef) as I suspect this would need to be fixed upstream.

CEF has it's own issue tracker at https://bitbucket.org/chromiumembedded/cef/issues?status=new&status=open

It's probably worth trying to disable CEF's network interceptor via the `--disable-request-handling-for-testing` command line arg to see if you get a different result.

> I have also attempted using the AddCrossOriginWhitelistEntry() method still to no avail (although I expected that not to work as I have seen that this method needs to be corrected in another discussion).

This is also a CEF issue see https://bitbucket.org/chromiumembedded/cef/issues/2918/cefaddcrossoriginwhitelistentry-not-fully

It's unlikely this can be fixed directly in `CefSharp`.

**Reply by smillsoper** [reply]
Thanks Alex, I'll see what disabling the network interceptor does and will definitely post this on CEF issue tracker as well.

**Reply by smillsoper** [reply]
I took this to the CEF forum as you suggested and they suggested I run through the same process in the Sample Application.

I found I received no CORS errors at all in the sample application and the page worked as expected.

I have since created a new application, added CefSharp.Winforms through NuGet and am still having the same CORS issue.  It seems that the issue is in CefSharp

**Reply by amaitland** [reply]
> I took this to the CEF forum as you suggested and they suggested I run through the same process in the Sample Application.

In future please test with the CEF Sample application before creating a new issue.

> I found I received no CORS errors at all in the sample application and the page worked as expected.

What version did you test with? 

What command line args did you test with? 

> I have since created a new application, added CefSharp.Winforms through NuGet and am still having the same CORS issue. It seems that the issue is in CefSharp

Does the problem reproduce with the minimum example? 

https://github.com/cefsharp/CefSharp.MinimalExample

**Reply by smillsoper** [reply]
Hi Alex,

Sorry - yes, I will test with the CEF Sample application first going forward.

I just tested with the minimum example with the same results.

The command line arguments that I have attempted thus far are:
disable-request-handling-for-testing
disable-web-security
disable-site-isolation-trials
disable-features=IsolateOrigins

All tests were performed with the following versions:
v99, v105, and v106 (all retrieved from NuGet)

Again, testing succeeded using the CEF Sample Application

**Reply by amaitland** [reply]
> Again, testing succeeded using the CEF Sample Application

Did you test with the following command line args? 

```
 cefclient.exe --multi-threaded-message-loop --no-sandbox
```

**Reply by smillsoper** [reply]
I just tested cefclient with the command line parameters you specified and the test still succeeds with cefclient.

**Reply by smillsoper** [reply]
Hi Alex, Just checking in, I haven't seen a reply on this thread in a couple of days.

**Reply by smillsoper** [reply]
the cef sample application version was also v106 - sorry I forgot to answer that question in my previous response

**Reply by amaitland** [reply]
@smillsoper Thanks for the detailed testing 👍 Can you confirm that your company is currently a [sponsor](https://github.com/sponsors/amaitland) or has a support arrangement in place? If yes please let me know which company and I'll happily prioritise looking into this.

**Reply by smillsoper** [reply]
Hi Alex I am now a sponsor. Glad to help out. CefSharp has become a major part of our call center software so was a no brainer for me. Thank you!

**Reply by amaitland** [reply]
Thanks 👍 

After quite a bit of digging it looks like the behaviour you are seeing is deliberate on behalf of the website, it looks like they are detecting that `CefSharp` is being used and altering behaviour.

Try renaming or disabling the object `CefSharp` creates in `javascript` (pick one).

```c#
//Rename object that CefSharp creates from cefSharp to notCefSharp (use your own name if you choose this option)
browser.JavascriptObjectRepository.Settings.JavascriptBindingApiGlobalObjectName = "notCefSharp";

//Disable creation of window.cefSharp and window.CefSharp
browser.JavascriptObjectRepository.Settings.JavascriptBindingApiEnabled = false;
```

If the problem still persists after making this change then try deleting the cache and retesting.

**Reply by smillsoper** [reply]
Thanks @amaitland, I'll give that a try.

**Reply by smillsoper** [reply]
@amaitland Fantastic! changing browser.JavascriptObjectRepository.Settings.JavascriptBindingApiGlobalObjectName did the trick! You're an amazing human being.

**Comment by n303ef** [comment]


**Reply by smillsoper** [reply]
Thanks for that, I'll give that a try

**Reply by smillsoper** [reply]
These command line parameters had no effect, thanks though.

**Reply by amaitland** [reply]
Disabling CORS and site isolation trials are not the same thing.
